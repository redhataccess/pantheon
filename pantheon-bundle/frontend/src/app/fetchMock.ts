if (!self.fetch) {
    self.fetch = (input: RequestInfo, init?: RequestInit | undefined) => {
        const p = new Promise<Response>((resolve, reject) => {
            // just don't do anything. this exists so that our tests don't try and actually resolve fetches,
            // (which is exactly what isomorphic-fetch tried to do while we were importing it...)
            // and calling nothing here allows us to skip all of our fetch.then() and fetch.catch() code.
        })
        return p
    }
}