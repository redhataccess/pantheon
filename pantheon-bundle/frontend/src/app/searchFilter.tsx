import React, { Component } from 'react';
import {
  Button, ButtonVariant, TextInput, InputGroup, Chip, ChipGroup, ChipGroupToolbarItem, FormSelect, FormSelectOption
} from '@patternfly/react-core';
import '@app/app.css';
import { SearchIcon, SortAlphaDownIcon, SortAlphaUpIcon } from '@patternfly/react-icons';

class SearchFilter extends Component<any, any> {
  constructor(props) {
    super(props);
    this.state = {
      allProducts: [],
      chipGroups: [],
      isSortedUp: true,
      moduleTypeValue: '',
      productOptions: [
        { value: '', label: 'Select a Product', disabled: false },
      ],
      productValue: '',
      productsQueryParam: '',
      productsToQuery: [],
      productsUUID: [],
      productversionsQueryParam: '',
      searchText: '',
      sortByValue: '',
      versionOptions: [
        { value: '', label: 'Select a Version', disabled: false },
      ],
      versionSelected: '',
      versionUUID: '',
      versionValue: '',
      versionsToQuery: [],
    };
  }

  public componentDidMount() {
    this.fetchProductVersionDetails()
  }

  public render() {
    const { chipGroups } = this.state;

    let verOptions = this.state.versionOptions
    if (this.state.allProducts[this.state.productValue]) {
      verOptions = this.state.allProducts[this.state.productValue]
    }

    const moduleTypeItems = [
      { value: 'All', label: 'All', disabled: false },
      { value: 'CONCEPT', label: 'Concept', disabled: false },
      { value: 'PROCEDURE', label: 'Procedure', disabled: false },
      { value: 'REFERENCE', label: 'Reference', disabled: false }
    ]

    const sortItems = [
      { value: 'Uploaded date', label: 'Uploaded date', disabled: false },
      { value: 'Title', label: 'Title', disabled: false },
      { value: 'Updated date', label: 'Updated date', disabled: false },
      { value: 'Module type', label: 'Module type', disabled: false }
    ]


    return (
      <React.Fragment>
        <div className="row-filter" >
          <InputGroup className="small-margin">
            <TextInput id="searchFilterInput" type="text" onKeyDown={this.props.onKeyDown} value={this.state.searchText} onChange={this.setSearchText} />
            <Button onClick={this.props.onClick} variant={ButtonVariant.control} aria-label="search button for search input">
              <SearchIcon />
            </Button>
          </InputGroup>


          <FormSelect value={this.state.productValue} onChange={this.onChangeProduct} aria-label="FormSelect Product" id="productForm">
            {this.state.productOptions.map((option, index) => (
              <FormSelectOption isDisabled={option.disabled} key={index} value={option.value} label={option.label} />
            ))}
          </FormSelect>

          <FormSelect className="small-margin" value={this.state.versionUUID} onChange={this.onChangeVersion} aria-label="FormSelect Version" id="productVersionForm">
            {verOptions.map((option) => (
              <FormSelectOption isDisabled={false} key={option.value} value={option.value} label={option.label} required={false} />
            ))}
          </FormSelect>

          <FormSelect className="small-margin" value={this.state.moduleTypeValue} onChange={this.onChangeModuleType} aria-label="FormSelect ModuleType" id="moduleTypeForm">
            {moduleTypeItems.map((option) => (
              <FormSelectOption isDisabled={false} key={option.value} value={option.value} label={option.label} required={false} />
            ))}
          </FormSelect>

          <FormSelect className="small-margin" value={this.state.sortByValue} onChange={this.onChangeSort} aria-label="FormSelect Sort" id="sortForm">
            {sortItems.map((option) => (
              <FormSelectOption isDisabled={false} key={option.value} value={option.value} label={option.label} required={false} />
            ))}
          </FormSelect>

          <Button onClick={this.setSortedUp} variant={ButtonVariant.control} aria-label="search button for search input">
            {this.state.isSortedUp ? <SortAlphaUpIcon /> : <SortAlphaDownIcon />}
          </Button>


        </div>
        <ChipGroup withToolbar={true}>
          {chipGroups.map(currentGroup => (
            <ChipGroupToolbarItem key={currentGroup.category} categoryName={currentGroup.category}>
              {currentGroup.chips.map(chip => (
                <Chip key={chip} onClick={this.deleteItem(currentGroup.category, chip)}>
                  {chip}
                </Chip>
              ))}
            </ChipGroupToolbarItem>
          ))}
        </ChipGroup>
      </React.Fragment>

    );
  }

  private setSearchText = (event) => this.setState({ searchText: event }, () => {
    this.setQuery();
  });

  private setSortedUp = () => {
    this.setState({ isSortedUp: !this.state.isSortedUp }, () => {
      this.setQuery()
    })
  };

  private fetchProductVersionDetails = () => {

    const path = '/content/products.3.json'
    let key
    const products = new Array()
    const prodUUID = new Array()

    fetch(path)
      .then((response) => {
        if (response.ok) {
          return response.json();
        } else if (response.status === 404) {
          return products
        } else {
          throw new Error(response.statusText);
        }
      })
      .then(responseJSON => {
        for (const i of Object.keys(responseJSON)) {
          key = i
          const nameKey = "name"
          const versionKey = "versions"
          const uuidKey = "jcr:uuid";
          if ((key !== 'jcr:primaryType')) {
            if (responseJSON[key][nameKey] !== undefined) {
              const pName = responseJSON[key][nameKey]
              const versionObj = responseJSON[key][versionKey]
              const productUUID = responseJSON[key][uuidKey]
              if (versionObj) {
                let vKey;
                const versions = [{ value: '', label: 'Select a Version', disabled: false }, { value: 'All', label: 'All', disabled: false },]
                for (const item in Object.keys(versionObj)) {
                  if (Object.keys(versionObj)[item] !== undefined) {
                    vKey = Object.keys(versionObj)[item]
                    if (vKey !== 'jcr:primaryType') {
                      if (versionObj[vKey][nameKey]) {
                        versions.push({ value: versionObj[vKey][uuidKey], label: versionObj[vKey][nameKey], disabled: false })
                      }
                    }
                  }
                }
                products[pName] = versions
                prodUUID[pName] = productUUID
              }
            }
          }
        }
        this.setState({
          allProducts: products,
          productsUUID: prodUUID
        })

        if (products) {
          const productItems = [{ value: 'Select a Product', label: 'Select a Product', disabled: false },]
          // tslint:disable-next-line: forin
          for (const item in products) {
            productItems.push({ value: item, label: item, disabled: false })
          }
          if (productItems.length > 1) {
            this.setState({ productOptions: productItems })
          }
        }
      })
      .catch((error) => {
        console.log(error)
      });
    return products;
  };

  private onChangeProduct = (productValue) => {
    this.setState({ productValue });
  }
  private onChangeVersion = () => {
    if (event !== undefined) {
      if (event.target !== null) {
        const selectedStr = "selectedOptions"
        if (this.state.versionUUID !== event.target[selectedStr][0].value) {
          this.setState({
            versionSelected: event.target[selectedStr][0].label,
            versionUUID: event.target[selectedStr][0].value,
            versionValue: event.target[selectedStr][0].label,
          }, () => {
            this.addChipItem();
          });
        }
      }
    }
  }

  private onChangeSort = (sortByValue) => {
    this.setState({ sortByValue }, () => {
      this.setQuery();
    });
  }
  private onChangeModuleType = (moduleTypeValue) => {
    this.setState({ moduleTypeValue }, () => {
      this.setQuery();
    });
  }

  private deleteItem = (product, id) => (event: any) => {
    const copyOfChipGroups = this.state.chipGroups;
    for (let i = 0; copyOfChipGroups.length > i; i++) {
      const category = copyOfChipGroups[i].category
      if (category === product) {
        const index = copyOfChipGroups[i].chips.indexOf(id);
        if (index !== -1) {
          const categoryKey = "category"
          product = copyOfChipGroups[i][categoryKey]
          copyOfChipGroups[i].chips.splice(index, 1);
          // check if this is the last item in the group category
          if (copyOfChipGroups[i].chips.length === 0) {
            copyOfChipGroups.splice(i, 1);
          }
        }
        break
      }
    }

    const uuidKey = "value"
    const versionUUID = this.state.allProducts[product].filter((e) => e.label === id)[0][uuidKey]
    const productUUID = this.state.productsUUID[product]
    if (versionUUID.trim() === "All") {
      let prodQuery = this.state.productsQueryParam
      prodQuery = prodQuery.replace("product=" + productUUID, '')
      if (prodQuery === '&') {
        prodQuery = ''
      }
      if (prodQuery.includes("&&")) {
        prodQuery = prodQuery.replace('&&', '&')
      }
      if (prodQuery.startsWith('&')) {
        prodQuery = prodQuery.substr(1)
      }
      this.setState({ chipGroups: copyOfChipGroups, productsQueryParam: prodQuery }, () => {
        this.setQuery();
      });
    } else {
      let verQuery = this.state.productversionsQueryParam
      verQuery = verQuery.replace("productversion=" + versionUUID, '')
      if (verQuery === '&') {
        verQuery = ''
      }
      if (verQuery.includes("&&")) {
        verQuery = verQuery.replace('&&', '&')
      }
      if (verQuery.startsWith('&')) {
        verQuery = verQuery.substr(1)
      }
      this.setState({ chipGroups: copyOfChipGroups, productversionsQueryParam: verQuery }, () => {
        this.setQuery();
      });
    }
  };

  private addChipItem = () => {
    const copyOfChipGroups = this.state.chipGroups;
    let exist = false
    let index = 0
    for (let i = 0; copyOfChipGroups.length > i; i++) {
      const category = copyOfChipGroups[i].category
      if (category === this.state.productValue) {
        exist = true
        index = i
        break
      }
    }
    if (exist) {
      let chipExists = false
      for (const i of copyOfChipGroups[index].chips) {
        const chip = i
        if (chip === this.state.versionSelected) {
          chipExists = true
        }

      }
      if (!chipExists && this.state.versionSelected.trim() !== "Select a Version") {
        copyOfChipGroups[index].chips.push(this.state.versionSelected);
      }
    } else {
      copyOfChipGroups.push({
        category: this.state.productValue,
        chips: [this.state.versionSelected]
      })
    }

    const uuidKey = "value"
    const versionUUID = this.state.allProducts[this.state.productValue].filter((e) => e.label === this.state.versionValue)[0][uuidKey]
    // If version is All just add the product.
    let prodQuery = this.state.productsQueryParam
    let verQuery = this.state.productversionsQueryParam
    if (versionUUID.trim() === "All") {
      if (this.state.productsQueryParam.trim() !== "") {
        prodQuery += '&'
      }
      if (!prodQuery.includes(this.state.productsUUID[this.state.productValue])) {
        prodQuery += "product=" + this.state.productsUUID[this.state.productValue]
      }
    } else if (versionUUID.trim() !== "") {
      if (this.state.productversionsQueryParam.trim() !== "") {
        verQuery += '&'
      }
      if (!verQuery.includes(versionUUID)) {
        verQuery += "productversion=" + versionUUID
      }
    }
    this.setState({ chipGroups: copyOfChipGroups, productsQueryParam: prodQuery, productversionsQueryParam: verQuery, productValue: '', versionSelected: '', versionUUID: '', versionValue: '' }, () => {
      this.setQuery();
    });
  };

  // Should be called after each change of state
  private setQuery = () => {
    let searchQuery = ""
    if (this.state.searchText.trim() !== "") {
      searchQuery += "search=" + this.state.searchText.trim()
    }

    if (this.state.productsQueryParam.trim() !== "") {
      if (searchQuery.trim() !== "") {
        searchQuery += "&"
      }
      searchQuery += this.state.productsQueryParam
    }

    if (this.state.productversionsQueryParam.trim() !== "") {
      if (searchQuery.trim() !== "") {
        searchQuery += "&"
      }
      searchQuery += this.state.productversionsQueryParam
    }

    // Default is All and should not add to the filter.
    if (this.state.moduleTypeValue.trim() !== "" && this.state.moduleTypeValue.trim() !== "All") {
      if (searchQuery.trim() !== "") {
        searchQuery += "&"
      }
      searchQuery += "type=" + this.state.moduleTypeValue
    }

    // Default key is Uploaded
    if (searchQuery.trim() !== "") {
      searchQuery += "&"
    }
    if (this.state.sortByValue.trim() === "") {
      searchQuery += "key=Uploaded"
    } else {
      searchQuery += "key=" + this.state.sortByValue
    }

    // isSortedUp is a boolean and will always have a set default
    if (searchQuery.trim() !== "") {
      searchQuery += "&"
    }
    searchQuery += "direction=" + (this.state.isSortedUp ? "desc" : "asc")

    if (searchQuery.includes("&&")) {
      searchQuery = searchQuery.replace('&&', '&')
    }

    this.props.filterQuery(searchQuery)
    console.log("This is the query: " + searchQuery)
  }
}

export { SearchFilter }; 