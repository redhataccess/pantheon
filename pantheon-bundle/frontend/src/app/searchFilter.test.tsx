import React from 'react';
import { SearchFilter } from '@app/searchFilter';
import { shallow } from 'enzyme';
import sinon from 'sinon'
import renderer from 'react-test-renderer'
import '@app/fetchMock'

const props = {
  filterQuery: () => ("any"),
}

describe('SearchFilter tests', () => {
  test('should render default Search component', () => {
    const view = shallow(<SearchFilter />);
    expect(view).toMatchSnapshot();
  });


  it('test fetchProductVersionDetails function', () => {
    const wrapper = renderer.create(<SearchFilter />);
    const inst = wrapper.getInstance()
    const spy = sinon.spy(inst, 'fetchProductVersionDetails')
    inst.componentDidMount()
    sinon.assert.called(spy)
  })

  it('test setQuery function', () => {
    const wrapper = renderer.create(<SearchFilter {...props} />)
    const inst = wrapper.getInstance()
    const spy = sinon.spy(inst, 'setQuery')
    inst.setQuery()
    sinon.assert.called(spy)
    inst.setState({ productsQueryParam: "stuff&"})
  })

  it('test onChangeProduct function', () => {
    const wrapper = renderer.create(<SearchFilter {...props} />)
    const inst = wrapper.getInstance()
    const spy = sinon.spy(inst, 'onChangeProduct')
    inst.onChangeProduct("prod")
    sinon.assert.called(spy)
  })

  it('test onChangeSort function', () => {
    const wrapper = renderer.create(<SearchFilter {...props} />)
    const inst = wrapper.getInstance()
    const spy = sinon.spy(inst, 'onChangeSort')
    inst.onChangeSort("asc")
    sinon.assert.called(spy)
  })


  it('test onChangeContentType function', () => {
    const wrapper = renderer.create(<SearchFilter {...props} />)
    const inst = wrapper.getInstance()
    const spy = sinon.spy(inst, 'onChangeContentType')
    inst.onChangeContentType("type")
    sinon.assert.called(spy)
  })

  it('test deleteItem function', () => {
    const wrapper = renderer.create(<SearchFilter {...props} />)
    const inst = wrapper.getInstance()
    const spy = sinon.spy(inst, 'deleteItem')

    const products = new Array()
    const pName = "productName";
    const versions = [{ value: '', label: 'Select a Version', disabled: false }, { value: 'All', label: 'All', disabled: false },];
    products[pName] = versions;
    inst.setState({ productValue: "productName", allProducts: products, chipGroups: [], })

    inst.deleteItem()
    sinon.assert.called(spy)
  })

  it('test addChipItem function', () => {
    const wrapper = renderer.create(<SearchFilter {...props} />)
    const inst = wrapper.getInstance()
    const spy = sinon.spy(inst, 'addChipItem')

    const products = new Array()
    const pName = "productName";
    const versions = [{ value: 'x', label: 'x', disabled: false }, { value: 'All', label: 'All', disabled: false },];
    products[pName] = versions;
    inst.setState({ versionValue:"x",productValue: "productName", allProducts: products, chipGroups: [], })

    inst.addChipItem()
    sinon.assert.called(spy)
  })
});
