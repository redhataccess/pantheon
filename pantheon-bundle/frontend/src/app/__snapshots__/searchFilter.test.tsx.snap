// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchFilter tests should render default Search component 1`] = `
<Fragment>
  <div
    className="row-filter"
  >
    <InputGroup
      className="small-margin"
    >
      <TextInput
        aria-label={null}
        className=""
        id="searchFilterInput"
        isDisabled={false}
        isReadOnly={false}
        isRequired={false}
        isValid={true}
        onChange={[Function]}
        type="text"
      />
      <Component
        aria-label="search button for search input"
        variant="control"
      >
        <SearchIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          title={null}
        />
      </Component>
    </InputGroup>
    <Component
      aria-label="search button for search input"
      variant="control"
    >
      <SortAlphaUpIcon
        color="currentColor"
        noVerticalAlign={false}
        size="sm"
        title={null}
      />
    </Component>
  </div>
</Fragment>
`;
